### PR TITLE
[WIP] Add coreclr_preload_assembly to CoreCLR host API

### PR DESCRIPTION
**Pull request by gbalykov**
This pull request adds `coreclr_preload_assembly` to host API, which loads assembly image to memory and applies relocations. `coreclr_preload_assembly` is supposed to be called only before initialization of CoreCLR.

Is it correct to call `PAL_malloc` before CoreCLR is initialized?

`PAL_ForceRelocs` is ignored for preloaded assembly images.

cc @ruben-ayrapetyan @Dmitri-Botcharnikov @kvochko

### CODE REVIEW

**Comment by jkotas** [comment]
#13936

cc @russellhadley loader change

File: src/pal/src/init/pal.cpp [hunk 1]
```diff
@@ -545,6 +545,7 @@ Initialize(
             ERROR("Unable to initialize virtual memory support\n");
             goto CLEANUP10;
         }
+        _ASSERTE(GetVirtualPageSize() == getpagesize());
```
Comment (ruben-ayrapetyan) [inline src/pal/src/init/pal.cpp]: This assertion is also added to `MAPMapPEFile`, so it seems redundant here.

File: src/pal/src/loader/module.cpp [hunk 2]
```diff
@@ -728,18 +728,19 @@ PAL_UnregisterModule(
 
 Parameters:
     IN hFile - file to map
+    IN prevLoadedBase - previously loaded base
```
Comment (ruben-ayrapetyan) [inline src/pal/src/loader/module.cpp]: Is it possible to set this argument to non-`NULL` in a scenario other than "preload ...  finish load"?
If not, it is better to add a comment that would indicate this.

It would be even better to have a comment with more detailed description of the preloading scheme, and to reference the comment from here and other places like this.

File: src/pal/src/map/map.cpp [hunk 3]
```diff
@@ -2138,6 +2138,27 @@ MAPRecordMapping(
     return palError;
 }
 
+static off_t
+MAPcalcAdj(
```
Comment (ruben-ayrapetyan) [inline src/pal/src/map/map.cpp]: Without looking into the function's code, it is not clear what it calculates. A brief comment (like near the neighbor functions) would be good.

File: src/pal/src/map/map.cpp [hunk 4]
```diff
@@ -2185,6 +2206,64 @@ MAPmmapAndRecord(
...
+
+    if (size1 == 0)
+    {
+        namePathString.CloseBuffer(0);
+        return NULL;
+    }
+
+    namePathString.CloseBuffer(size1 - 1);
+
+    return name;
```
Comment (ruben-ayrapetyan) [inline src/pal/src/map/map.cpp]: Is it valid after `namePathString` is destructed upon function exit?

File: src/pal/src/map/map.cpp [hunk 5]
```diff
@@ -2193,96 +2272,170 @@ MAPmmapAndRecord(
...
 
-    ENTRY("MAPMapPEFile (hFile=%p)\n", hFile);
+    size_t headerSize;
 
-    //Step 0: Verify values, find internal pal data structures.
-    if (INVALID_HANDLE_VALUE == hFile)
+    void * loadedBase = prevLoadedBase;
+    bool isPreloaded = loadedBase != NULL;
+    if (!((doPreload && prevLoadedBase == NULL && hFile == NULL && path != NULL && size != NULL)
+          || (!doPreload && hFile != NULL && path == NULL && size == NULL)))
```
Comment (ruben-ayrapetyan) [inline src/pal/src/map/map.cpp]: It would be great to split calculation of the expression into several steps with named intermediate values (like `isPreloaded`)

File: src/pal/src/map/map.cpp [hunk 6]
```diff
@@ -2193,96 +2272,170 @@ MAPmmapAndRecord(
...
     if (sizeof(ntHeader) != read(fd, &ntHeader, sizeof(ntHeader)))
     {
-        palError = FILEGetLastErrorFromErrno();
+        if (!doPreload)
+        {
+            palError = FILEGetLastErrorFromErrno();
+        }
+        else
+        {
+            palError = ERROR_INVALID_NAME;
```
Comment (ruben-ayrapetyan) [inline src/pal/src/map/map.cpp]: This case seems to be unrelated to a name or path. What do you think about using a more general error code here?

File: src/pal/src/map/map.cpp [hunk 7]
```diff
@@ -2303,27 +2456,33 @@ void * MAPMapPEFile(HANDLE hFile)
         || (preferredBase + virtualSize < preferredBase)    // Does the image overflow?
         )
     {
-        ERROR_(LOADER)( "image is corrupt\n" );
+        if (!doPreload)
+        {
+            ERROR_(LOADER)( "image is corrupt\n" );
```
Comment (ruben-ayrapetyan) [inline src/pal/src/map/map.cpp]: What do you think about possibility to move this into a helper macro, that would take two arguments - the message and the condition?

File: src/pal/src/map/map.cpp [hunk 8]
```diff
@@ -2303,27 +2456,33 @@ void * MAPMapPEFile(HANDLE hFile)
...
         goto done;
     }
 
 #if _DEBUG
-    envVar = EnvironGetenv("PAL_ForceRelocs");
-    if (envVar)
+    if (!doPreload)
     {
-        if (strlen(envVar) > 0)
+        envVar = EnvironGetenv("PAL_ForceRelocs");
```
Comment (ruben-ayrapetyan) [inline src/pal/src/map/map.cpp]: Is the value ignored in case of preloading?

File: src/vm/corhost.cpp [hunk 9]
```diff
@@ -416,6 +416,43 @@ void SetCommandLineArgs(LPCWSTR pwzAssemblyPath, int argc, LPCWSTR* argv)
     GCPROTECT_END();
 }
 
+HRESULT CorHost2::PreloadAssembly(LPCWSTR wszAssemblyPath)
+{
+#ifndef FEATURE_PAL
+    return E_FAIL;
```
Comment (ruben-ayrapetyan) [inline src/vm/corhost.cpp]: Please, add a comment, which describes why this case is not supported.

File: src/dlls/mscoree/unixinterface.cpp [hunk 10]
```diff
@@ -146,6 +146,37 @@ extern "C" int coreclr_create_delegate(void*, unsigned int, const char*, const c
...
+int coreclr_preload_assembly(
+            const char *assemblyPath
+)
+{
+    if (assemblyPath == NULL)
+    {
+        return E_FAIL;
+    }
+
+    int length = MultiByteToWideChar(CP_UTF8, 0, assemblyPath, -1, NULL, 0);
```
Comment (janvorli) [inline src/dlls/mscoree/unixinterface.cpp]: Please use StringToUnicode to make the conversion instead.

File: src/dlls/mscoree/unixinterface.cpp [hunk 11]
```diff
@@ -146,6 +146,37 @@ extern "C" int coreclr_create_delegate(void*, unsigned int, const char*, const c
...
+    ASSERTE_ALL_BUILDS(length != 0);
+
+    LPWSTR result = (LPWSTR) PAL_malloc(sizeof(WCHAR) * length);
+    ASSERTE_ALL_BUILDS(result != NULL);
+
+    length = MultiByteToWideChar(CP_UTF8, 0, assemblyPath, -1, result, length);
+    ASSERTE_ALL_BUILDS(length != 0);
+
+    return CorHost2::PreloadAssembly(result);
```
Comment (janvorli) [inline src/dlls/mscoree/unixinterface.cpp]: It would be preferable to let the PreloadAssembly make a copy of the string instead of passing ownership to match other CorHost2 functions. Could you please change that? Then use ConstWStringHolder here to store the converted string so that it gets automatically destroyed when the coreclr_preload_assembly returns.

File: src/inc/corhost.h [hunk 12]
```diff
@@ -389,6 +389,11 @@ class CorHost2 :
         LPCWSTR* argv,
         DWORD* pReturnValue);
 
+    static HRESULT PreloadAssembly(
+        LPCWSTR pwzAssemblyPath);
```
Comment (janvorli) [inline src/inc/corhost.h]: A nit - could you please match the naming of other string arguments of methods in this class? That means naming this wszAssemblyPath.

File: src/vm/peimage.h [hunk 13]
```diff
@@ -372,12 +372,32 @@ class PEImage
...
+#endif
 
 public:
     void CachePEKindAndMachine();
     void GetPEKindAndMachine(DWORD* pdwKind, DWORD* pdwMachine);
 
+public:
+
+#ifdef FEATURE_PAL
+    static void addNewPreloadedAssembly(LPCWSTR path, PEImageLayout *layout)
```
Comment (janvorli) [inline src/vm/peimage.h]: Could you please follow the coding convention of the surrounding code and use pascal casing of the method names?

File: src/vm/peimagelayout.cpp [hunk 14]
```diff
@@ -85,6 +102,21 @@ PEImageLayout* PEImageLayout::Map(HANDLE hFile, PEImage* pOwner)
     RETURN pAlloc.Extract();    
 }
 
+#ifdef FEATURE_PAL
+PEImageLayout* PEImageLayout::MapPreload(LPCWSTR path)
+{
+    void *addr = PAL_malloc(sizeof(MinimalMappedImageLayout));
```
Comment (janvorli) [inline src/vm/peimagelayout.cpp]: Why do you use malloc and placement new instead of just regular `new` everywhere?

File: src/vm/peimagelayout.cpp [hunk 15]
```diff
@@ -363,11 +399,65 @@ ConvertedImageLayout::ConvertedImageLayout(PEImageLayout* source)
 
 #ifdef CROSSGEN_COMPILE
     if (HasNativeHeader())
-        ApplyBaseRelocations();
+        ApplyBaseRelocations(false, false);
 #endif
 }
 
-MappedImageLayout::MappedImageLayout(HANDLE hFile, PEImage* pOwner)
+MinimalMappedImageLayout::MinimalMappedImageLayout(LPCWSTR assemblyPath)
```
Comment (janvorli) [inline src/vm/peimagelayout.cpp]: What is the rationale behind naming this layout "Minimal"? Can you please explain that?

File: src/dlls/mscoree/unixinterface.cpp (hunk 10 above)
Comment (gbalykov) [reply src/dlls/mscoree/unixinterface.cpp]: StringToUnicode cannot be used here directly because of `new`, which can't be used before CoreCLR is initialized. I have updated pull request description with some details

File: src/vm/peimagelayout.cpp (hunk 14 above)
Comment (gbalykov) [reply src/vm/peimagelayout.cpp]: `new` can't be used before CoreCLR is initialized, and MapPreload is supposed to be called before CoreCLR is initialized. I have updated pull request description with some details

File: src/vm/peimagelayout.cpp (hunk 15 above)
Comment (gbalykov) [reply src/vm/peimagelayout.cpp]: `MinimalMappedImageLayout` holds pointer to mapped assembly, which will be used later to initialize `MappedImageLayout`. As `MinimalMappedImageLayout` is only used for this, I thought it was good to name it as some kind of minimal version of `MappedImageLayout`. I think it could also be called `PreloadedMappedImageLayout` or something like that. What do you think about this?

File: src/vm/peimagelayout.cpp (hunk 14 above)
Comment (janvorli) [reply src/vm/peimagelayout.cpp]: You don't need to have coreclr initialized to use PAL functions, you just need to initialize PAL in the coreclr_preload_assembly. Would that cause trouble with the forking? 
Anyways, while calling PAL_malloc works now, calling PAL functions without initializing the PAL is not a good thing. We might decide to add some logging or something that depends on the PAL initialization to that function later.

File: src/vm/peimagelayout.cpp (hunk 15 above)
Comment (janvorli) [reply src/vm/peimagelayout.cpp]: That would be a great name for that.

File: src/vm/peimagelayout.cpp (hunk 14 above)
Comment (gbalykov) [reply src/vm/peimagelayout.cpp]: We considered initialization of PAL before coreclr_initialize, but it creates `CPalSynchMgrController`, which prevents forking after preload. Which functions does this thread currently perform? Could it be created later or replaced with something else?

Comment (janvorli) [reply src/vm/peimagelayout.cpp]: This thread manages inter-process synchronization. It seems that we could create and start the thread separately from the other initialization. A reasonable way to do that could be to:
* Add a new PAL initialization method called PAL_InitializeStartThreads. This method would be called after the fork.
* Move the code from inside of `if (flags & PAL_INITIALIZE_SYNC_THREAD)` in the Initialization of PAL to the new method
* In the new method, ensure that the thread starting code could executed only once in a process even if the function is called multiple times, similar to what we do in the Initialize function.
* In the new method, ensure that proper cleanup happens if starting the thread fails.
* Remove the PAL_INITIALIZE_SYNC_THREAD constant definition and usage from pal.h

Comment (jkotas) [reply src/vm/peimagelayout.cpp]: Note that there are number of other places that are not fork-compatible - both in the PAL initialization and CoreCLR initialization. One example from many: `gPID = getpid();`. So this is not just about starting threads.

Comment (janvorli) [reply src/vm/peimagelayout.cpp]: Ah, right.

Comment (gbalykov) [reply src/vm/peimagelayout.cpp]: What is a minimal amount of initialization that should happen for PAL in order to use some functions from it like PAL_malloc?

If there is a strict border between two stages of PAL initialization, assert could be added to all PAL function with the condition that appropriate stage of PAL initialization has completed.

For example, there is one more place that creates thread - `SEHInitializeMachExceptions` call in `Initialize` from PAL. Could everything before it be moved to the first stage, and everything after - to the second? Then, PAL_malloc will check that first stage has completed.

There is also a concern that PEImage is used without initialization of CoreCLR. Could the parts related to preloading be moved from CoreCLR and PAL to the separate component, which will be initialized before preloading, and will initialize minimum from PAL?

Comment (jkotas) [reply src/vm/peimagelayout.cpp]: I think it would be best to move the list of pre-loaded images to the PAL. Once it is in the PAL you have much better control over what it calls, and it would require initializing the PAL at all in the ideal case.

Comment (gbalykov) [reply src/vm/peimagelayout.cpp]: This is a good idea, we tried to think in this direction, but the problem is that non-PAL version seems to use `MappedImageLayout` too, and we couldn't simply move `MappedImageLayout` to PAL. Is it required for non-PAL version?

Comment (gbalykov) [reply src/vm/peimagelayout.cpp]: To clarify, should LoadLibrary load native image and apply relocations, or MappedImageLayout can be used for this?

Comment (jkotas) [reply src/vm/peimagelayout.cpp]: I would avoid touching MappedImageLayout.

Instead:
- Refactor MAPMapPEFile to part that just does the memory mapping and the rest
- Create a clone of `PEImageLayout::ApplyBaseRelocations` in the PAL that works directly against the metal without, and does not need any of the PAL initialized
- Change PAL_LOADLoadPEFile (or introduce a new API) to consult the list of preloaded files first and return it if there is one.

**Comment by gbalykov** [comment]
@dotnet-bot test Tizen armel Cross Debug Build

**Comment by jkotas** [comment]
@gbalykov Do you plan to finish this?

**Comment by jkotas** [comment]
Closing inactive PR. Feel free to submit a new PR if you get a chance to finish this.

**Comment by gbalykov** [comment]
@jkotas We have currently frozen this activity. I'll open another PR, as you said, if we continue it, thank you.
